File: golang_cli/crypto/crypto_test.go
// crypto_test.go
package crypto

import (
	"bytes"
	"crypto/aes"
	"crypto/rand"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"io"
	"math/big"
	"os"
	"testing"

	"github.com/ethereum/go-ethereum/common"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func encryptDecrypt(t *testing.T, plaintextBytes []byte) []byte {
	key := make([]byte, aes.BlockSize)
	_, err := rand.Read(key)
	if err != nil {
		t.Fatalf("Failed to generate random key: %v", err)
	}

	ciphertext, r, err := Encrypt(key, plaintextBytes)
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}

	decrypted, err := Decrypt(key, r, ciphertext)
	if err != nil {
		t.Fatalf("Decrypt failed: %v", err)
	}

	return decrypted
}

func TestEncryptDecrypt(t *testing.T) {
	// Arrange
	// Create plaintext with the value 100 as a big integer with 128 bits
	plaintextValue := big.NewInt(100)
	plaintextBytes := make([]byte, aes.BlockSize)
	plaintextValue.FillBytes(plaintextBytes)

	// Act
	decrypted := encryptDecrypt(t, plaintextBytes)

	// Assert
	assert.Equal(t, plaintextBytes, decrypted, "Decrypted message should match the original plaintext")
}

func TestEncryptDecryptWithPadding(t *testing.T) {
	// Arrange
	// Create plaintext with the value 100 as a big integer with less than 128 bits
	plaintextValue := big.NewInt(100)
	plaintextBytes := plaintextValue.Bytes()

	// Act
	decrypted := encryptDecrypt(t, plaintextBytes)
	decryptedValue := new(big.Int).SetBytes(decrypted)

	// Assert
	assert.Equal(t, plaintextValue, decryptedValue, "Decrypted message should match the original plaintext")
}

func readEncryptionFromFile(path string) ([]byte, []byte, []byte, error) {
	data, err := readValFromFile(path)
	if err != nil {
		return nil, nil, nil, err
	}

	// Split the data into two hex strings
	hexStrings := bytes.Split(data, []byte("\n"))

	// Convert the hex strings to bytes
	if len(hexStrings) != 3 {
		return nil, nil, nil, fmt.Errorf("Expected three hex strings in the file")
	}

	keyBytes, err := hex.DecodeString(string(hexStrings[0]))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, nil, nil, err
	}

	cipherBytes, err := hex.DecodeString(string(hexStrings[1]))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, nil, nil, err
	}

	randomBytes, err := hex.DecodeString(string(hexStrings[2]))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, nil, nil, err
	}

	return keyBytes, cipherBytes, randomBytes, nil
}

func checkEncryption(t *testing.T, filePath string) {
	// Arrange
	plaintextValue := big.NewInt(100)

	// Read encryption from a file
	key, ciphertext, r, err := readEncryptionFromFile(filePath)
	require.NoError(t, err, "Read encryption should not return an error")

	err = os.Remove(filePath)
	require.NoError(t, err, "Delete file should not return an error")

	// Act
	decrypted, err := Decrypt(key, r, ciphertext)
	require.NoError(t, err, "Decrypt should not return an error")

	decryptedValue := new(big.Int).SetBytes(decrypted)

	// Assert
	assert.Equal(t, plaintextValue, decryptedValue, "Python decrypted message should match the original plaintext")
}

func TestPythonJSEnsryption(t *testing.T) {

	checkEncryption(t, "../../python/soda_python_sdk/test_pythonEncryption.txt")
	checkEncryption(t, "../../js/test_jsEncryption.txt")
	checkEncryption(t, "../../ts/test_tsEncryption.txt")
}

func TestLoadWriteAESKey(t *testing.T) {
	// Arrange and Assert
	// Create a temporary file for testing
	tempFile := "temp_key_file.txt"
	defer os.Remove(tempFile)

	key := make([]byte, aes.BlockSize)
	_, err := rand.Read(key)
	if err != nil {
		t.Fatalf("Failed to generate random key: %v", err)
	}

	// Act and Assert
	// Test WriteAESKey
	err = WriteAESKey(tempFile, key)
	require.NoError(t, err, "WriteAESKey should not return an error")

	// Test LoadAESKey
	loadedKey, err := LoadAESKey(tempFile)
	require.NoError(t, err, "LoadAESKey should not return an error")
	assert.Equal(t, key, loadedKey, "Loaded key should match the original key")
}

func TestGenerateAndWriteAESKey(t *testing.T) {
	// Arrange and Assert
	// Create a temporary file for testing
	tempFile := "temp_key_file.txt"
	defer os.Remove(tempFile)
	// Test GenerateAESKey and WriteAESKey
	key, err := GenerateAESKey()
	require.NoError(t, err, "GenerateAESKey should not return an error")

	// Act and Assert
	err = WriteAESKey(tempFile, key)
	require.NoError(t, err, "WriteAESKey should not return an error")

	// Verify that the key was written to the file
	loadedKey, err := LoadAESKey(tempFile)
	require.NoError(t, err, "LoadAESKey should not return an error")
	assert.Equal(t, key, loadedKey, "Loaded key should match the generated key")
}

func TestSignature(t *testing.T) {
	// Arrange
	sender := make([]byte, AddressSize)
	_, err := rand.Read(sender)
	addr := make([]byte, AddressSize)
	_, err = rand.Read(addr)
	funcSig := make([]byte, FuncSigSize)
	_, err = rand.Read(funcSig)

	key := GenerateECDSAPrivateKey()
	require.NoError(t, err, "Failed to generate random key")

	// Create plaintext with the value 100 as a big integer with less than 128 bits
	plaintextValue := big.NewInt(100)
	plaintextBytes := plaintextValue.Bytes()
	ciphertext, r, err := Encrypt(key, plaintextBytes)
	require.NoError(t, err, "Encrypt should not return an error")

	ct := append(ciphertext, r...)

	// Act and assert
	signature, err := SignIT(sender, addr, funcSig, ct, key)
	require.NoError(t, err, "Sign should not return an error")

	// Verify the signature
	verified := VerifyIT(sender, addr, funcSig, ct, signature)

	assert.Equal(t, verified, true, "Verify signature should return true")
}

func readValFromFile(path string) ([]byte, error) {
	// Open the file for reading
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("Error:", err)
		return nil, err
	}
	defer file.Close() // Make sure to close the file when done

	// Read the contents of the file as a string
	data, err := io.ReadAll(file)
	if err != nil {
		fmt.Println("Error:", err)
		return nil, err
	}

	return data, nil
}

func readHexValFromFile(path string) ([]byte, error) {
	data, err := readValFromFile(path)
	if err != nil {
		return nil, err
	}

	// Convert the hex string to bytes
	hexBytes, err := hex.DecodeString(string(data))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, err
	}

	return hexBytes, nil
}

func readSigFromFileAndCompare(t *testing.T, filePath string, signature []byte) {
	fileSig, err := readHexValFromFile(filePath)
	require.NoError(t, err, "Read Signature should not return an error")

	err = os.Remove(filePath)
	require.NoError(t, err, "Delete file should not return an error")

	assert.Equal(t, fileSig, signature, "signature should match the python signature")
}

// TestFixedMsgSignature is a test function that checks the functionality of the Sign and VerifySignature functions.
// It first decodes fixed hexadecimal strings into byte slices representing the sender, address, function signature, ciphertext, and key.
// It then signs a message using these parameters and checks for errors.
// The function then reads signatures from two files (one Python and one JavaScript) and compares them to the generated signature.
// This ensures that the signature is correct in both python and javascript implementations.
// It then verifies the signature using the VerifySignature function and asserts that the result should be true.
func TestFixedMsgSignature(t *testing.T) {
	// Arrange
	// Create plaintext with the value 100 as a big integer with less than 128 bits
	sender, _ := hex.DecodeString("d67fe7792f18fbd663e29818334a050240887c28")
	addr, _ := hex.DecodeString("69413851f025306dbe12c48ff2225016fc5bbe1b")
	funcSig, _ := hex.DecodeString("dc85563d")
	ct, _ := hex.DecodeString("f8765e191e03bf341c1422e0899d092674fc73beb624845199cd6e14b7895882")
	key, _ := hex.DecodeString("3840f44be5805af188e9b42dda56eb99eefc88d7a6db751017ff16d0c5f8143e")

	// Act and assert
	// Sign the message
	signature, err := SignIT(sender, addr, funcSig, ct, key)
	require.NoError(t, err, "Sign should not return an error")

	// Reading from file simulates the communication between the evm (golang) and the user (python/js)
	readSigFromFileAndCompare(t, "../../python/soda_python_sdk/test_pythonSignature.txt", signature)
	readSigFromFileAndCompare(t, "../../js/test_jsSignature.txt", signature)
	readSigFromFileAndCompare(t, "../../ts/test_tsSignature.txt", signature)

	// Verify the signature
	verified := VerifyIT(sender, addr, funcSig, ct, signature)

	assert.Equal(t, verified, true, "Verify signature should return true")
}

func TestIT(t *testing.T) {
	// Arrange
	// Create plaintext with the value 100 as a big integer with less than 128 bits
	plaintext := uint64(100)
	sender := common.HexToAddress("d67fe7792f18fbd663e29818334a050240887c28")
	contract := common.HexToAddress("69413851f025306dbe12c48ff2225016fc5bbe1b")
	funcSig := "test(bytes)"
	userKey, _ := hex.DecodeString("b3c3fe73c1bb91862b166a29fe1d63e9")
	signingKey, _ := hex.DecodeString("3840f44be5805af188e9b42dda56eb99eefc88d7a6db751017ff16d0c5f8143e")

	// Act and assert
	// Sign the message
	ct, signature, err := prepareIT(plaintext, userKey, sender, contract, funcSig, signingKey)
	require.NoError(t, err, "Sign should not return an error")

	plaintextBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(plaintextBytes, plaintext)

	checkIT(t, plaintextBytes, userKey, sender.Bytes(), contract.Bytes(), GetFuncSig(funcSig), ct.Bytes(), signature)

	// Reading from file simulates the communication between the evm (golang) and the user (python/js)
	pythonCt, pythonSignature, err := readTwoHexStringsFromFile("../../python/soda_python_sdk/test_pythonIT.txt")
	require.NoError(t, err, "Read file should not return an error")
	checkIT(t, plaintextBytes, userKey, contract.Bytes(), GetFuncSig(funcSig), ct.Bytes(), pythonCt, pythonSignature)
	err = os.Remove("../../python/test_pythonIT.txt")

	jsCt, jsSignature, err := readTwoHexStringsFromFile("../../js/test_jsIT.txt")
	tsCt, tsSignature, err := readTwoHexStringsFromFile("../../ts/test_tsIT.txt")

	require.NoError(t, err, "Read file should not return an error")
	checkIT(t, plaintextBytes, userKey, contract.Bytes(), GetFuncSig(funcSig), ct.Bytes(), jsCt, jsSignature)
	checkIT(t, plaintextBytes, userKey, contract.Bytes(), GetFuncSig(funcSig), ct.Bytes(), tsCt, tsSignature)
	err = os.Remove("../../js/test_jsIT.txt")
	err = os.Remove("../../ts/test_tsIT.txt")
}

func checkIT(t *testing.T, plaintext, userKey, sender, addr, funcSig, ct, signature []byte) {
	// Verify the signature
	verified := VerifyIT(sender, addr, funcSig, ct, signature)
	assert.Equal(t, verified, true, "Verify signature should return true")

	decryptedText, err := Decrypt(userKey, ct[aes.BlockSize:], ct[:aes.BlockSize])
	if err != nil {
		t.Fatalf("Decrypt failed: %v", err)
	}

	// Verify decrypted plaintext matches original message
	assert.Equal(t, plaintext, decryptedText[len(decryptedText)-len(plaintext):], "Decrypted plaintext should match original message")
}

func TestRSAEncryption(t *testing.T) {
	// Arrange
	// Generate key pair
	privateKey, publicKey, err := GenerateRSAKeyPair()
	require.NoError(t, err, "Generate RSA key pair should not return an error")

	// Message to encrypt
	plaintext := []byte("hello world")

	// Act and assert
	// Encrypt the plaintext
	cipher, err := EncryptRSA(publicKey, plaintext)
	require.NoError(t, err, "Encrypt should not return an error")

	// Decrypt the ciphertext
	decryptedText, err := DecryptRSA(privateKey, cipher)
	require.NoError(t, err, "Decrypt should not return an error")

	// Verify decrypted plaintext matches original message
	assert.Equal(t, plaintext, decryptedText, "Decrypted plaintext should match original message")
}

func readTwoHexStringsFromFile(path string) ([]byte, []byte, error) {
	data, err := readValFromFile(path)
	if err != nil {
		return nil, nil, err
	}

	// Split the data into two hex strings
	hexStrings := bytes.Split(data, []byte("\n"))

	// Convert the hex strings to bytes
	if len(hexStrings) != 2 {
		return nil, nil, fmt.Errorf("Expected two hex strings in the file")
	}

	privateKeyBytes, err := hex.DecodeString(string(hexStrings[0]))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, nil, err
	}

	publicKeyBytes, err := hex.DecodeString(string(hexStrings[1]))
	if err != nil {
		fmt.Println("Error decoding hex:", err)
		return nil, nil, err
	}

	return privateKeyBytes, publicKeyBytes, nil
}

func appendHexToFile(filename string, hexString string) error {
	// Check if the file exists
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		// Create the file if it doesn't exist
		_, err := os.Create(filename)
		if err != nil {
			return err
		}
	}

	// Open the file for appending with write permissions
	file, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer file.Close()

	// Write the hexadecimal string to the file
	_, err = fmt.Fprintf(file, "%s\n", hexString)
	if err != nil {
		return err
	}

	return nil
}

// encryptMessage is a test helper function that reads RSA keys from a file, encrypts a plaintext message using the RSA public key,
// and then appends the encrypted message (in hexadecimal format) to the same file.
// It takes two parameters:
// t: The testing object, used for reporting errors in the test execution.
// keysFilePath: The path to the file containing the RSA keys.
// The function first reads the RSA keys from the file and checks for errors.
// It then defines a plaintext message to be encrypted.
// The plaintext message is encrypted using the RSA public key, and the function checks for errors.
// Finally, the encrypted message is converted to hexadecimal format and appended to the file containing the RSA keys.
func encryptMessage(t *testing.T, keysFilePath string) {
	// Reading and writing from/to a file simulates the communication between the evm (golang) and the user (python/js)
	_, key, err := readTwoHexStringsFromFile(keysFilePath)
	require.NoError(t, err, "Read RSA keys should not return an error")

	// Message to encrypt
	plaintext := []byte("hello world")

	// Decrypt the ciphertext
	ciphertext, err := EncryptRSA(key, plaintext)
	require.NoError(t, err, "Encrypt should not return an error")

	appendHexToFile(keysFilePath, "\n"+hex.EncodeToString(ciphertext))
}

// TestRSAEncryptionFixed is a test function that encrypts a fixed message using the RSA public keys generated by python and javascript tests.
// It uses the helper function encryptMessage to encrypt a plaintext message using RSA keys stored in files.
// The encrypted message is then appended to the same file.
// After the encryption is appended to the same file, the python and javascript tests check the encrypted message against the expected value.
// This test simulates the case that a user generates his own RSA keys in python or javascript tools and after that decrypts a message that was encrypted in the evm (which uses Go).
func TestRSAEncryptionFixed(t *testing.T) {
	encryptMessage(t, "../../python/soda_python_sdk/test_pythonRSAEncryption.txt")
	encryptMessage(t, "../../js/test_jsRSAEncryption.txt")
	encryptMessage(t, "../../ts/test_tsRSAEncryption.txt")
}

func checkFunctionSignature(t *testing.T, filePath string, expected []byte) {
	val, err := readHexValFromFile(filePath)
	require.NoError(t, err, "Read python value should not return an error")
	assert.Equal(t, expected, val, "hashed values should match")
	err = os.Remove(filePath)
}

func TestGetFuncSig(t *testing.T) {
	functionSig := "sign(bytes)"
	// Hash the function signature using Keccak-256 and return the first 4 bytes
	hash := GetFuncSig(functionSig)

	// Check that the python hashed value matches the Golang hashed value
	filePath := "../../python/soda_python_sdk/test_pythonFunctionKeccak.txt"
	checkFunctionSignature(t, filePath, hash)

	// Check that the js hashed value matches the Golang hashed value
	filePath = "../../js/test_jsFunctionKeccak.txt"
	checkFunctionSignature(t, filePath, hash)

}

File: ts/crypto.test.ts
import {
    ADDRESS_SIZE,
    BLOCK_SIZE,
    decrypt, decryptRSA,
    encrypt, encryptRSA,
    generateAesKey,
    generateECDSAPrivateKey, generateRSAKeyPair, getFuncSig, HEX_BASE, prepareIT, prepareMessage, signIT, prepareIT256, writeBigUInt256BE, CT_SIZE,
    verifySignatures, extractSignatureComponents,
    loadAesKey, writeAesKey, signEIP712, buildOnboardUserTypedData, buildEncryptToUserTypedData, recoverAddressFromEIP712Signature
} from "./crypto"
import fs from 'fs';
import crypto from 'crypto';
import {
    Address,
    ecrecover,
    hashPersonalMessage,
    pubToAddress,
    privateToPublic,
    toBuffer,
    toChecksumAddress
} from "ethereumjs-util";
import {ethers} from "ethers";
import * as assert from "node:assert";

function uint8ArrayToBigInt(uint8Array: Uint8Array): bigint {
    let value = BigInt(0);
    for (let i = 0; i < uint8Array.length; i++) {
        value = (value << 8n) | BigInt(uint8Array[i]);
    }
    return value;
}

describe('Crypto Tests', () => {

    // Test case for encrypt and decrypt
    it('should encrypt and decrypt successfully', () => {
        // Arrange
        const plaintextInteger = 100;
        const plaintextBuffer = Buffer.alloc(1);
        plaintextBuffer.writeUInt8(plaintextInteger);

        // Act
        const key = generateAesKey();

        const { ciphertext, r } = encrypt(key, plaintextBuffer);

        const decryptedBuffer = decrypt(key, r, ciphertext);

        // Write Buffer to file to later check in Go
        fs.writeFileSync("test_tsEncryption.txt", key.toString('hex') + "\n" + ciphertext.toString('hex') + "\n" + r.toString('hex'));

        const decryptedInteger =  uint8ArrayToBigInt(decryptedBuffer)

        // Assert
        assert.strictEqual(decryptedInteger, BigInt(plaintextInteger));
    });

    // Test case for load and write AES key
    it('should load and write AES key successfully', () => {
        // Arrange
        const key = generateAesKey();

        // Act
        writeAesKey('key.txt', key);
        const loadedKey = loadAesKey('key.txt');

        // Assert
        assert.deepStrictEqual(loadedKey, key);

        // Delete the key file
        fs.unlink('key.txt', (err: NodeJS.ErrnoException | null) => {
            if (err) {
                console.error('Error deleting file:', err);
            }
        });

    });

    // Test case for invalid plaintext size
    it('should throw error for invalid plaintext size', () => {
        // Arrange
        const key = generateAesKey();
        const plaintextBuffer = Buffer.alloc(20); // Bigger than 128 bits

        // Act and Assert
         assert.throws(() => encrypt(key, plaintextBuffer), RangeError);

    });

    // Test case for invalid ciphertext size
    it('should throw error for invalid ciphertext size', () => {
        // Arrange
        const key = generateAesKey();
        const ciphertext = Buffer.from([0x01, 0x02, 0x03]); // Smaller than 128 bits
        const r = Buffer.alloc(BLOCK_SIZE);

        // Act and Assert
        assert.throws(() => decrypt(key, r, ciphertext), RangeError);
    });

    // Test case for invalid random size
    it('should throw error for invalid random size', () => {
        // Arrange
        const key = generateAesKey();
        const r = Buffer.from([0x01, 0x02, 0x03]); // Smaller than 128 bits
        const ciphertext = Buffer.alloc(BLOCK_SIZE);

        // Act and Assert
        assert.throws(() => decrypt(key, r, ciphertext), RangeError);
    });

    // Test case for invalid key size
    it('should throw error for invalid key size', () => {
        // Arrange
        const key = Buffer.from([0x01, 0x02, 0x03]); // Smaller than 128 bits

        // Act and Assert
        // Test invalid key size when writing key
        assert.throws(() => writeAesKey('key.txt', key), RangeError);

        // Test invalid key size when encrypting
        const plaintextBuffer = Buffer.alloc(BLOCK_SIZE);
        assert.throws(() => encrypt(key, plaintextBuffer), RangeError);

        // Test invalid key size when decrypting
        const ciphertext = Buffer.alloc(BLOCK_SIZE);
        const r = Buffer.alloc(BLOCK_SIZE);
        assert.throws(() => decrypt(key, r, ciphertext), RangeError);
    });

    // Test case for verify signature
    it('should sign and verify the signature', () => {
        // Arrange
        // Simulate the generation of random bytes
        const sender = crypto.randomBytes(ADDRESS_SIZE);
        const addr = crypto.randomBytes(ADDRESS_SIZE);
        let key = generateECDSAPrivateKey();

        // Create a ciphertext
        const plaintextBuffer = Buffer.alloc(1);
        plaintextBuffer.writeUInt8(100);
        const aeskey = generateAesKey();
        const { ciphertext, r } = encrypt(aeskey, plaintextBuffer);
        let ct = Buffer.concat([ciphertext, r]);

        // Act
        // Generate the signature
        const signatureBytes = signIT(sender, addr, ct, key);

        const publicKey = privateToPublic(key);
        const signerAddress = toChecksumAddress('0x' + pubToAddress(publicKey).toString('hex'));
        const message = Buffer.concat([sender, addr, ct]);
        const verified = verifySignatures(message, [signatureBytes], [signerAddress]);

        // Assert
        assert.strictEqual(verified, true);
    });

    // Test case for verify signature
    it('should sign and verify the EIP191 signature', () => {
        // Arrange
        // Simulate the generation of random bytes
        const sender = crypto.randomBytes(ADDRESS_SIZE);
        const addr = crypto.randomBytes(ADDRESS_SIZE);
        let key = generateECDSAPrivateKey();

        // Create a ciphertext
        const plaintextBuffer = Buffer.alloc(1);
        plaintextBuffer.writeUInt8(100);
        const aeskey = generateAesKey();
        const { ciphertext, r } = encrypt(aeskey, plaintextBuffer);
        let ct = Buffer.concat([ciphertext, r]);

        // Act
        // Generate the signature
        const signatureBytes = signIT(sender, addr, ct, key, true);

        const {rBytes, sBytes, vByte} = extractSignatureComponents(signatureBytes);

        // Verify the signature
        const expectedPublicKey = privateToPublic(key);
        const expectedAddress = toChecksumAddress('0x' + expectedPublicKey.toString('hex'));

        const message = Buffer.concat([sender, addr, ct]);
        const hash = hashPersonalMessage(message);

        // Recover the public key from the signature
        const publicKey = ecrecover(hash, vByte, rBytes, sBytes);
        // Derive the Ethereum address from the recovered public key
        const address = toChecksumAddress('0x' + publicKey.toString('hex'));

        // Compare the derived address with the expected signer's address
        const isVerified = address === expectedAddress;

        // Assert
        assert.strictEqual(isVerified, true);
    });

    it('should sign and recover the EIP712 onboard signature', async () => {
        const { publicKey } = generateRSAKeyPair();
        const key = generateECDSAPrivateKey();
        const wallet = new ethers.Wallet(ethers.hexlify(key));
        const address = Buffer.from(wallet.address.slice(2), 'hex');

        const typedData = buildOnboardUserTypedData(publicKey, address, 0);
        const signature = await signEIP712(typedData, key);
        const recoveredAddress = recoverAddressFromEIP712Signature(typedData, signature);

        assert.strictEqual(recoveredAddress.toLowerCase(), wallet.address.toLowerCase());
    });

    it('should sign and recover the EIP712 encrypt-to-user signature', async () => {
        const key = generateECDSAPrivateKey();
        const wallet = new ethers.Wallet(ethers.hexlify(key));
        const handle = Buffer.from('81ff8a56f19f4ffd576e57a01f3c0f256de80517a4e4385470d1c33fe7804fe7', 'hex');

        const typedData = buildEncryptToUserTypedData([handle], null, 11155111);
        const signature = await signEIP712(typedData, key);
        const recoveredAddress = recoverAddressFromEIP712Signature(typedData, signature);

        assert.strictEqual(recoveredAddress.toLowerCase(), wallet.address.toLowerCase());
    });

    // Test case for verify signature
    it('should sign a fixed message and write the signature to a file', () => {
        // Arrange
        // Simulate the generation of random bytes
        const sender = Buffer.from('8f01160c98e5cdfa625197849c85cf5fc1f76b1b', 'hex');
        const addr = Buffer.from('69413851f025306dbe12c48ff2225016fc5bbe1b', 'hex');
        const ct = Buffer.from('81ff8a56f19f4ffd576e57a01f3c0f256de80517a4e4385470d1c33fe7804fe7', 'hex');
        const key = Buffer.from('3840f44be5805af188e9b42dda56eb99eefc88d7a6db751017ff16d0c5f8143e', 'hex');

        // Act
        // Generate the signature
        const signature = signIT(sender, addr, ct, key);

        const filename = 'test_tsSignature.txt'; // Name of the file to write to

        // Convert hexadecimal string to buffer
        let sigString = signature.toString('hex');

        // Write buffer to the file, this simulates the communication between the evm (golang) and the user (python/js)
        fs.writeFile(filename, sigString, (err) => {
            if (err) {
                console.error('Error writing to file:', err);
                return;
            }
        });
    });

    it('should prepareMessage using fixed data', async () => {
        // Arrange
        // Simulate the generation of random bytes
        const plaintext = BigInt("100");
        const userKey = 'b3c3fe73c1bb91862b166a29fe1d63e9';
        const senderAddress ='0x8f01160c98e5cdfa625197849c85cf5fc1f76b1b';
        const contractAddress = '0x69413851f025306dbe12c48ff2225016fc5bbe1b';
        const signingKey = '0x3840f44be5805af188e9b42dda56eb99eefc88d7a6db751017ff16d0c5f8143e';

        // Act
        // Generate the signature
        const {
            message
        } = prepareMessage(plaintext, senderAddress, userKey, contractAddress);

        const wallet = new ethers.Wallet(signingKey);
        const signature = await wallet.signMessage(message);

        const recoveredAddress = ethers.verifyMessage(message, signature);

        assert.strictEqual(recoveredAddress.toLowerCase(), senderAddress.toLowerCase());
    });


    // Test case for verify signature
    it('should prepare IT using fixed data', () => {
        // Arrange
        // Simulate the generation of random bytes
        const plaintext = BigInt("100");
        const userKey = Buffer.from('b3c3fe73c1bb91862b166a29fe1d63e9', 'hex');;
        const sender = new Address(toBuffer(Buffer.from('8f01160c98e5cdfa625197849c85cf5fc1f76b1b', 'hex')));
       
        // Act
        // Generate the signature
        const {userAddress, ctInt} = prepareIT(plaintext, userKey, sender.toBuffer());

        const ctHex = ctInt.toString(HEX_BASE).padStart(CT_SIZE * 2, '0');
        // Create a Buffer to hold the bytes (CT_SIZE = 32 bytes = 2 * BLOCK_SIZE)
        const ctBuffer = Buffer.from(ctHex, 'hex');

        // Write Buffer to file to later check in Go
        fs.writeFileSync("test_tsIT.txt", ctHex);

        // Decrypt the ct and check the decrypted value is equal to the plaintext
        const decryptedBuffer = decrypt(userKey, ctBuffer.subarray(BLOCK_SIZE, ctBuffer.length), ctBuffer.subarray(0, BLOCK_SIZE));

        // Convert the plaintext to bytes
        const hexString = plaintext.toString(16);
        const plaintextBytes = Buffer.from(hexString, 'hex');
        // Assert
        const expectedBytes = decryptedBuffer.subarray(decryptedBuffer.length - plaintextBytes.length, decryptedBuffer.length)
        assert.deepStrictEqual(plaintextBytes.toString('hex'), Buffer.from(expectedBytes).toString('hex'));
        const intResult = uint8ArrayToBigInt(decryptedBuffer);
        assert.deepStrictEqual(plaintext, intResult);
    });

    // Test case for verify signature
    it('should prepare IT 256 bits using fixed data', () => {
        // Arrange
        // Simulate the generation of random bytes
        const plaintext = BigInt("34028236692093846346337460743176821145600");
        const userKey = Buffer.from('b3c3fe73c1bb91862b166a29fe1d63e9', 'hex');;
        const sender = new Address(toBuffer(Buffer.from('8f01160c98e5cdfa625197849c85cf5fc1f76b1b', 'hex')));
        
        // Act
        // Generate the signature
        const {userAddress, ciphertext} = prepareIT256(plaintext, userKey, sender.toBuffer());

        const ctHighBytes = Buffer.alloc(CT_SIZE); // Allocate a buffer of size 32 bytes
        writeBigUInt256BE(ctHighBytes, ciphertext.ciphertextHigh); // Write the uint256 value to the buffer as big-endian
        const ctLowBytes = Buffer.alloc(CT_SIZE); // Allocate a buffer of size 32 bytes
        writeBigUInt256BE(ctLowBytes, ciphertext.ciphertextLow); // Write the uint256 value to the buffer as big-endian
        // Decrypt the ct and check the decrypted value is equal to the plaintext
        const decryptedBuffer = decrypt(userKey, ctHighBytes.subarray(BLOCK_SIZE, ctHighBytes.length), ctHighBytes.subarray(0, BLOCK_SIZE), ctLowBytes.subarray(BLOCK_SIZE, ctLowBytes.length), ctLowBytes.subarray(0, BLOCK_SIZE));

        // Convert the plaintext to bytes
        const hexString = plaintext.toString(16).padStart(64, '0');
        const plaintextBytes = Buffer.from(hexString, 'hex'); 

        // Assert
        const expectedBytes = decryptedBuffer.subarray(decryptedBuffer.length - plaintextBytes.length, decryptedBuffer.length)
        assert.deepStrictEqual(plaintextBytes.toString('hex'), Buffer.from(expectedBytes).toString('hex'));
        const intResult = uint8ArrayToBigInt(decryptedBuffer);
        assert.deepStrictEqual(plaintext, intResult);
    });

    // Test case for test rsa encryption scheme
    it('should encrypt and decrypt a message using RSA scheme', () => {
        // Arrange
        const plaintext ='hello world';
        const plaintextBuffer = Buffer.from(plaintext);

        const { publicKey, privateKey } = generateRSAKeyPair();

        // Act
        const ciphertext = encryptRSA(publicKey, plaintext);

        const hexString = privateKey.toString('hex') + "\n" + publicKey.toString('hex') + "\n" + Buffer.from(ciphertext).toString('hex');

        // Write buffer to the file
        const filename = 'test_tsRSAEncryption.txt'; // Name of the file to write to
        fs.writeFileSync(filename, hexString);

        const decrypted = decryptRSA(privateKey, Buffer.from(ciphertext).toString('hex'));

        // Assert
        assert.deepStrictEqual(plaintextBuffer, Buffer.from(decrypted));
    });

    function readHexFromFile(filename:string) {
        return new Promise((resolve, reject) => {
            fs.readFile(filename, 'utf8', (err, data) => {
                if (err) {
                    reject(err);
                    return;
                }

                const lines = data.trim().split('\n');
                if (lines.length >= 3) {
                    const hexData1 = lines[0].trim();
                    const hexData2 = lines[1].trim();
                    const hexData3 = lines[2].trim();
                    resolve([hexData1, hexData2, hexData3]);
                } else {
                    reject(new Error('Not enough lines in the file.'));
                }
            });
        });
    }

    // Test case for test rsa decryption scheme
    it('should decrypt a message using RSA scheme', async () => {
        // Arrange
        const plaintext = Buffer.from('hello world');

        // Act
        // Read private key and ciphertext
        // Reading from file simulates the communication between the evm (golang) and the user (python/js)
        try {
            const value = await readHexFromFile('test_tsRSAEncryption.txt');
            const [hexData1, hexData2, hexData3] = value as [string, string, string];
            const privateKey = Buffer.from(hexData1, 'hex');
            const ciphertext = Buffer.from(hexData3, 'hex').toString('hex');

            const decrypted = decryptRSA(privateKey, ciphertext);

            const decryptedBuffer = Buffer.from(decrypted);

            // Assert
            assert.deepStrictEqual(plaintext, decryptedBuffer);
            fs.unlinkSync('test_tsRSAEncryption.txt');
        } catch (error) {
            console.error("Error reading file:", error);
            throw error;
        }
    });

    // Test case for test function signature
    it('should hash a function signature', () => {
        // Arrange
        const functionSig = 'sign(bytes)';

        // Act
        const hash = getFuncSig(functionSig);

        const filename = 'test_tsFunctionKeccak.txt'; // Name of the file to write to
        // Write Buffer to file
        fs.writeFileSync(filename, hash.toString('hex'));

    });
});


